src/clegoues/genprog4java/Search/Search.java

```diff
@@ -263,41 +263,51 @@ public void mutate(Representation<G> variant) throws GiveUpException {
 
 		ArrayList<Location> faultyAtoms = localization.getFaultLocalization();
 		
-		ArrayList<Location> proMutList = new ArrayList<Location>();
 		boolean foundMutationThatCanApplyToAtom = false;
 		boolean alreadySetAllStmtsToFixLoc = false;
 		while(!foundMutationThatCanApplyToAtom){
+			ArrayList<Location> proMutList = new ArrayList<Location>();
 			//promut default is 1 // promut stands for proportional mutation rate, which controls the probability that a genome is mutated in the mutation step in terms of the number of genes within it should be modified.
 			for (int i = 0; i < Search.promut; i++) {
 				//chooses a random location
-				Pair<?, Double> wa = null;
-				boolean alreadyOnList = false;
+				Location wa = null;
+				boolean foundUsableLoc = false;
+
 				//If it already picked all the fix atoms from current FixLocalization, then start picking from the ones that remain
 				if(proMutList.size()>=faultyAtoms.size()){ 
 					localization.setAllPossibleStmtsToFixLocalization();				
 					alreadySetAllStmtsToFixLoc = true;
+					faultyAtoms = localization.getFaultLocalization();
 				}
-				//only adds the random atom if it is different from the others already added
-				do {
+
+				//only adds the random atom if it is different from the others already added & it has available mutations
+				while(!foundUsableLoc) {
 					//chooses a random faulty atom from the subset of faulty atoms
-					wa = localization.getRandomLocation(Configuration.randomizer.nextDouble());
-					// insert a check to see if this location has any valid mutations?  If not, look again
-					// if not, somehow tell the variant to remove that location from the list of faulty atoms
-					alreadyOnList = proMutList.contains(wa);
-				} while(alreadyOnList);
-				proMutList.add((Location)wa);
+					wa = (Location) GlobalUtils.chooseOneWeighted(new ArrayList(faultyAtoms), Configuration.randomizer.nextDouble());
```

Comment:
Significantly uglier; I just copied what was in `localization.getRandomLocation` and pasted it here except changed the list of choices to `faultyAtoms`; open to suggestions